client-java-core/src/main/java/io/featurehub/client/ApplyFeature.java

```diff
@@ -28,60 +28,58 @@ public ApplyFeature(PercentageCalculator percentageCalculator, MatcherRepository
-
-          int basePercentageVal = basePercentage.computeIfAbsent(newPercentageKey, (k) -> 0);
-          // if we have changed the key or we have never calculated it, calculate it and set the
-          // base percentage to null
-          if (percentage == null || !newPercentageKey.equals(percentageKey)) {
-            percentageKey = newPercentageKey;
-
-            percentage = percentageCalculator.determineClientPercentage(percentageKey,
-              featureValueId);
-            log.info("percentage for {} on {} calculated at {}", defaultPercentageKey, key, percentage);
-          }
+    if (cac == null || strategies == null || strategies.isEmpty()) {
+      return new Applied(false, null);
+    }
+    Integer percentage = null;
+    String percentageKey = null;
+    Map<String, Integer> basePercentage = new HashMap<>();
+    String defaultPercentageKey = cac.defaultPercentageKey();
+
+    for(FeatureRolloutStrategy rsi : strategies ) {
+      if (rsi.getPercentage() != null && (defaultPercentageKey != null || !rsi.getPercentageAttributes().isEmpty())) {
+        // determine what the percentage key is
+        String newPercentageKey = determinePercentageKey(cac, rsi.getPercentageAttributes());
+
+        int basePercentageVal = basePercentage.computeIfAbsent(newPercentageKey, (k) -> 0);
+        // if we have changed the key or we have never calculated it, calculate it and set the
+        // base percentage to null
+        if (percentage == null || !newPercentageKey.equals(percentageKey)) {
+          percentageKey = newPercentageKey;
+          percentage = percentageCalculator.determineClientPercentage(percentageKey, featureValueId);
+          log.info("percentage for {} on {} calculated at {}", defaultPercentageKey, key, percentage);
+        }
 
-          log.info("comparing actual {} vs required: {}", percentage, rsi.getPercentage());
-          int useBasePercentage = rsi.getAttributes() == null || rsi.getAttributes().isEmpty() ? basePercentageVal : 0;
-          // if the percentage is lower than the user's key +
-          // id of feature value then apply it
-          if (percentage <= (useBasePercentage + rsi.getPercentage())) {
-            if (rsi.getAttributes() != null && !rsi.getAttributes().isEmpty()) {
-              if (matchAttributes(cac, rsi)) {
-                return new Applied(true, rsi.getValue());
-              }
-            } else {
-              return new Applied(true, rsi.getValue());
-            }
+        log.info("comparing actual {} vs required: {}", percentage, rsi.getPercentage());
+        int useBasePercentage = hasAttributes(rsi) ? basePercentageVal : 0;
+        // if the percentage is lower than the user's key +
+        // id of feature value then apply it
+        if (percentage <= (useBasePercentage + rsi.getPercentage())) {
+          if (hasAttributes(rsi) || matchAttributes(cac, rsi)) {
+            return new Applied(true, rsi.getValue());
```

Comment:
this has changed the logic - if there are attributes and they match or if there are no attributes. So this would need to be !hasAttributes(rsi) || matchAttributes(cac,rsi))?

Reply:
ideally if the tests passed and i am reasoning correcting, we need another test!

Reply:
Sorry this is my mistake. This I did manually and hence screwed it up. This should be hasAttributes(rsi) **&&** matchAttributes(cac, rsi) as this was a nested if.

Reply:
But even if that were true, won't it miss the case where the accumulated percentage is less than the tagged percentage and there are no attributes to match? Thats an unconditional match at that percentage level, unless you simplified it elsewhere?